src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -2069,6 +2219,115 @@ void M(bool b)
+  IL_000c:  ret
+}
+");
+        }
+
+        [Fact, WorkItem(50096, "https://github.com/dotnet/roslyn/issues/50096")]
+        public void FunctionPointerInferenceInParameter_ImplicitReferenceConversionOnParameter()
+        {
+            var verifier = CompileAndVerify(@"
+unsafe
+{
+    Test(string.Empty, &converter);
+
+    static string converter(object o) => string.Empty;
+    static void Test<T1, T2>(T1 t1, delegate*<T1, T2> func) {}
+}
+", options: TestOptions.UnsafeReleaseExe, verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped);
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       17 (0x11)
+  .maxstack  2
+  IL_0000:  ldsfld     ""string string.Empty""
+  IL_0005:  ldftn      ""string <Program>$.<<Main>$>g__converter|0_0(object)""
+  IL_000b:  call       ""void <Program>$.<<Main>$>g__Test|0_1<string, string>(string, delegate*<string, string>)""
+  IL_0010:  ret
+}
+");
+        }
+
+        [Fact, WorkItem(50096, "https://github.com/dotnet/roslyn/issues/50096")]
+        public void FunctionPointerInferenceInReturn()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe
+{
+    Test(0, &converter);
+
+    static int converter(string v) => 0;
+    static void Test<T1, T2>(T2 t2, delegate*<T1, T2> func) {}
+}
+", options: TestOptions.UnsafeReleaseExe);
+
+            // It might seem like this should have the same behavior as FunctionPointerInferenceInParameter. However, this is not how method group resolution works.
+            // We don't look at return types when resolving a method group, which means that we can't use T2 to narrow down the candidates. We therefore can't
+            // find any information about T1, and resolution fails. FunctionPointerInferenceInParameter, on the other hand, has a bound for T1: int. That bound
+            // is then used as an input to method group resolution, and we can narrow down to string converter(int v). The return type of that method can then be
+            // used as a bound to infer T2.
+            comp.VerifyDiagnostics(
+                // (4,5): error CS0411: The type arguments for method 'Test<T1, T2>(T2, delegate*<T1, T2>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //     Test(0, &converter);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test").WithArguments("Test<T1, T2>(T2, delegate*<T1, T2>)").WithLocation(4, 5),
+                // (7,17): warning CS8321: The local function 'Test' is declared but never used
+                //     static void Test<T1, T2>(T2 t2, delegate*<T1, T2> func) {}
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "Test").WithArguments("Test").WithLocation(7, 17)
+            );
+        }
+
+        [Fact, WorkItem(50096, "https://github.com/dotnet/roslyn/issues/50096")]
+        public void FunctionPointerInference_ThroughMethodGroup()
```

Comment:
>FunctionPointerInference_ThroughMethodGroup [](start = 20, length = 43)

I am assuming this test is added to cover changes made in commit 4. It feels like we need to add similar coverage for an address of and lambdas.